org.eclipse.tm4e.core/src/main/java/org/eclipse/tm4e/core/internal/grammar/Grammar.java

```diff
@@ -225,8 +225,11 @@ public ITokenizeLineResult2 tokenizeLine2(String lineText, StackElement prevStat
 			isFirstLine = false;
 			prevState.reset();
 		}
-		
-		lineText = lineText + '\n';
+
+		if(lineText.isEmpty() || lineText.charAt(lineText.length()-1) != '\n') {
```

Comment:
Well, right now I'm using the grammar directly, and as I already had a StringBuffer with the contents, I always make sure it ends with a \n -- as it's always required, having clients add it is usually better (to avoid having to allocate a new copy just to add a '\n').